src/block.c

```diff
@@ -258,4 +259,1694 @@ coap_add_data_blocked_response(coap_resource_t *resource,
+
+        cq->observe_set = 0;
+        if (pdu == NULL)
+          return 0;
+        /* Need to make sure that this is the correct type */
+        pdu->type = type;
+
+        coap_update_option(pdu, COAP_OPTION_OBSERVE,
+                           coap_encode_var_safe(buf, sizeof(buf),
+                                                COAP_OBSERVE_CANCEL),
+                           buf);
+        mid = coap_send(session, pdu);
+        if (mid != COAP_INVALID_TID)
+          return 1;
+        break;
+      }
+    }
+  }
+  return 0;
+}
+
+int
+coap_add_data_large_internal(coap_session_t *session,
+                             coap_pdu_t *pdu,
+                             coap_resource_t *resource,
+                             const coap_string_t *query,
+                             int maxage,
+                             uint64_t etag,
+                             size_t length,
+                             const uint8_t *data,
+                             coap_release_large_data_t release_func,
+                             void *app_ptr) {
+
+  ssize_t avail;
+  coap_block_t block;
+  size_t chunk;
+  coap_lg_xmit_t *lg_xmit = NULL;
+  uint8_t buf[8];
+  int have_block_defined = 0;
+  uint8_t blk_size;
+  uint16_t option;
+
+  assert(pdu);
+
+  if (!(session->block_mode & COAP_BLOCK_USE_LIBCOAP)) {
+    coap_log(LOG_DEBUG,
+             "** %s: coap_send_large: COAP_BLOCK_USE_LIBCOAP not enabled\n",
+             coap_session_str(session));
+    goto add_data;
+  }
+
+/* Block NUM max 20 bits and block size is "2**(SZX + 4)" and SZX max of 6 */
+#define MAX_BLK_LEN (((1 << 20) - 1) * (1 << (6 + 4)))
+
+  if (length > MAX_BLK_LEN) {
+    coap_log(LOG_WARNING,
+             "Size of large buffer restricted to 0x%x bytes\n", MAX_BLK_LEN);
+    length = MAX_BLK_LEN;
+  }
+  /* Determine the block size to use, adding in sensible options if needed */
+  if (COAP_PDU_IS_REQUEST(pdu)) {
+    coap_lg_xmit_t *q = NULL;
+
+    option = COAP_OPTION_BLOCK1;
+
+    /* See if this token is already in use for large bodies (unlikely) */
+    lg_xmit = session->lg_xmit;
+    while (lg_xmit) {
+      if (full_match(pdu->token, pdu->token_length,
+                     lg_xmit->b.b1.app_token->s,
+                     lg_xmit->b.b1.app_token->length)) {
+        /* Unfortunately need to free this off as potential size change */
+        if (q)
+          q->next = lg_xmit->next;
+        else
+          session->lg_xmit = lg_xmit->next;
+        coap_block_delete_lg_xmit(session, lg_xmit);
+        break;
+      }
+      q = lg_xmit;
+      lg_xmit = lg_xmit->next;
+    }
+  }
+  else {
+    /* Have to assume that it is a response even if code is 0.00 */
+    coap_lg_xmit_t *q = NULL;
+    coap_string_t empty = { 0, NULL};
+
+    assert(resource);
+    option = COAP_OPTION_BLOCK2;
+
+    /* Check if resource+query is already in use for large bodies (unlikely) */
+    lg_xmit = session->lg_xmit;
+    while (lg_xmit) {
```

Comment:
I prefer to always use the utlist macros to improve readability and to avoid errors due to own linked-list implementations.

Reply:
OK - will work on using LL_